osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/configuration/service/GetConfigurationObjectServiceDsmr4.java

```diff
@@ -0,0 +1,118 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Component;
+
+@Component
+public class GetConfigurationObjectServiceDsmr4 extends GetConfigurationObjectService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(GetConfigurationObjectServiceDsmr4.class);
+
+    private static final int NUMBER_OF_CONFIGURATION_OBJECT_ELEMENTS = 2;
+    private static final int INDEX_OF_GPRS_OPERATION_MODE = 0;
+    private static final int INDEX_OF_CONFIGURATION_FLAGS = 1;
+
+    private final DlmsHelper dlmsHelper;
+
+    public GetConfigurationObjectServiceDsmr4(final DlmsHelper dlmsHelper) {
+        this.dlmsHelper = dlmsHelper;
+    }
+
+    @Override
+    public boolean handles(final Protocol protocol) {
+        return protocol == Protocol.DSMR_4_2_2;
+    }
+
+    @Override
+    ConfigurationObjectDto getConfigurationObject(final GetResult result) throws ProtocolAdapterException {
+        final DataObject resultData = result.getResultData();
+        if (resultData == null || !resultData.isComplex()) {
+            final String message = String.format("Expected ConfigurationObject ResultData as Complex, but got: %s",
+                    resultData);
+            LOGGER.warn(message);
+            throw new ProtocolAdapterException(message);
+        }
+        return this.getConfigurationObject(resultData);
+    }
+
+    private ConfigurationObjectDto getConfigurationObject(final DataObject resultData) throws ProtocolAdapterException {
+        LOGGER.info("ConfigurationObject ResultData: {}", this.dlmsHelper.getDebugInfo(resultData));
+        final List<DataObject> elements = resultData.getValue();
+        if (elements == null || elements.size() != NUMBER_OF_CONFIGURATION_OBJECT_ELEMENTS) {
+            final String message = String.format("Expected ConfigurationObject ResultData with %d elements, but got %s",
+                    NUMBER_OF_CONFIGURATION_OBJECT_ELEMENTS, elements == null ? "null" : elements.size());
+            LOGGER.warn(message);
+            throw new ProtocolAdapterException(message);
+        }
+        return this.getConfigurationObject(elements);
+    }
+
+    private ConfigurationObjectDto getConfigurationObject(final List<DataObject> elements)
+            throws ProtocolAdapterException {
+        final GprsOperationModeTypeDto gprsMode = this.getGprsOperationMode(elements.get(INDEX_OF_GPRS_OPERATION_MODE));
+        final ConfigurationFlagsDto flags = this.getConfigurationFlags(elements.get(INDEX_OF_CONFIGURATION_FLAGS));
+        return new ConfigurationObjectDto(gprsMode, flags);
+    }
+
+    private GprsOperationModeTypeDto getGprsOperationMode(final DataObject gprsMode) throws ProtocolAdapterException {
+        if (gprsMode == null || !gprsMode.isNumber()) {
+            final String message = String.format(
+                    "Expected ConfigurationObject gprsOperationMode as Number, but got: %s", gprsMode);
+            LOGGER.warn(message);
```

Comment:
a warning is logged although an error is thrown.

Reply:
I agree this is bad practice. It was even worse, the messages were not the same...

I could remove the logging but thought it might be handy to quickly diagnose problems on the playground. (I don't know where the ProtocolAdapterException and its message end up).

Reply:
you don't need to remove the logging, but changing warn to error would be better.

Reply:
Oh, I thought you were (correctly IMO) pointing out that it's a bad practice to log AND throw exceptions. I'd be happier to remove the logging (after this feature is tested successfully) than to promote it to ERROR level.